### PR TITLE
Added support for `logit_bias` in token generation [WIP / Looking for Feedback]

### LINKED ISSUE #240: [Feature] support for general logit processors
Are there any plans to support logit processors via an additional API parameters? For instance, the OpenAI API provides a `logit_bias` parameter, that is applied to the token distribution during generation.

This would be great for http://github.com/eth-sri/lmql, where we heavily rely on token masking to enforce constraints during decoding.

Thank you for your hard work on this project.

### PR DESCRIPTION
**Pull request by MarcusDunn**
# What does this PR do?

Adds support for sending `logit_bias`'s during token generation using [SequenceBiasLogitsProcessor](https://huggingface.co/docs/transformers/main/en/internal/generation_utils#transformers.SequenceBiasLogitsProcessor).

Fixes #240 

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests) Pull Request section?
- [ ] Did you make sure to update the documentation with your changes.
- [ ] Did you write any new necessary tests?

## Tasks still remaining.

- [ ] Fix `FlashCausalLMBatch.concatenate`

   As we need to tokenize the inputs to `logit_bias` in order to create a `SequenceBiasLogitsProcessor` there needs to be a  tokenizer passed into `NextTokenChooser.from_pb` and `HeterogeneousNextTokenChooser.from_pb` this was no issue everywhere but `FlashCausalLMBatch.concatenate`. 

    I'm looking for feedback on how to resolve this nicely - the obvious solution would be to store a `tokenizer` in `FlashCausalLMBatch` or `HeterogeneousNextTokenChooser` similar to how `device` is selected when concatenating at the moment.

- [ ] Add Python tests

- [ ] Update client

- [ ] Documentation

    Not 100% if this is complete. I've updated the rust openapi schema and added docs to the protobuf files - not sure if this is sufficient.

## Who can review?

@OlivierDehaene

### CODE REVIEW

File: server/text_generation_server/models/flash_causal_lm.py [hunk 1]
```diff
@@ -634,6 +634,7 @@ def concatenate(cls, batches: List["FlashCausalLMBatch"]) -> "FlashCausalLMBatch
             next_token_chooser_parameters,
             dtype=batches[0].next_token_chooser.dtype,
             device=batches[0].next_token_chooser.device,
+            tokenizer=batches[0].next_token_chooser.tokenizer
```
Comment (MarcusDunn) [inline server/text_generation_server/models/flash_causal_lm.py:637]: *here* is where I need access to a tokenizer but one was not conveniently already available in scope.

File: server/text_generation_server/utils/tokens.py [hunk 2]
```diff
@@ -218,12 +241,15 @@ def __init__(
         self.do_sample = do_sample
         self.dtype = dtype
         self.device = device
+        self.tokenizer = tokenizer
```
Comment (MarcusDunn) [inline server/text_generation_server/utils/tokens.py:244]: so unfortunately I had to add a field to `HeterogeneousNextTokenChooser`

File: server/text_generation_server/utils/tokens.py [hunk 3]
```diff
@@ -275,8 +302,12 @@ def from_pb(
             typical_p=[pb_.typical_p for pb_ in pb],
             do_sample=[pb_.do_sample for pb_ in pb],
             seeds=[pb_.seed for pb_ in pb],
+            logit_bias=[dict(
+                [(tuple(tokenizer.encode(bias.string, add_special_tokens=False).input_ids[0]), bias.bias) for bias in
+                 pb_.logit_bias]) for pb_ in pb],
```
Comment (MarcusDunn) [inline server/text_generation_server/utils/tokens.py:307]: my python is pretty bad - this felt really ugly, not certain how to improve it.

**Comment by MarcusDunn** [comment]
@Narsil this is probably ready for review (I understand @OlivierDehaene is off for August?).

It would be great if you could approve workflows, I cannot build the docker image for the life of me and cannot test CUDA stuff without it.

**Comment by Narsil** [comment]
Hi thanks for the PR.

This will likely wait for @OlivierDehaene return.
There are a lot of ways to push models in certain directions (guidance, token healing, speculative decoding, logits bias).

I'm relatively convinced, that modifying biases like you're suggesting is not desirable. This is something that should either be in the model itself, or access higher level of controls.

That being said, all methods mentioned above will likely need to find a common underlying arch (so we don't have to recreate everything all the time.)

Efficiency is also an important factor, adding new logits modifiers adds overhead for all clients on the machine.
Check how any modification can impact performance: https://github.com/huggingface/text-generation-inference/pull/617

**Comment by MarcusDunn** [comment]
We sometimes call out to a LLM to ask a true/false question (or select from a series of dynamically generated options) to make some decision. The context can get fairly long and it takes some time to prompt engineer and test our way into getting an expected answer consistently.

This prompt engineering process is repeated every time we swap out a model. (Falcon -> LLama 2 chat -> LLama 2 instruction-finetuned), and we're never 100% confident that our prompt is sufficient to guarantee the output format we want. It slows down testing the latest and greatest from the open source community.

I've seen guidance/lmql, it would 100% do what we need, but seems like gross overkill and I do not want to maintain a HTTP interface in front of it if I can avoid it. 

I would love a way to guide output, I can see that being totally out of scope for this project though, this felt like a reasonable medium.

**Comment by Narsil** [comment]
> what do you mean by "access higher level of controls"?

Typically guidance. You would specific something higher level that raw token_id (which most user do not have access to).
Also because specifying ids require specific knowledge of the model, which is something most users do not have. The prompt is already quite tricky for finetuned models.

**Comment by ggbetz** [comment]
Hello! @narsil wrote:

> Could you share maybe more information about the problem you're trying to solve ?

I'm working with [LMQL](https://github.com/eth-sri/lmql) and if I understand correctly, `logit_bias` support would be essential for using TGI - HF Inference Endpoints as an open cloud alternative to OpenAI. (See issue https://github.com/eth-sri/lmql/issues/190.)

**Comment by Narsil** [comment]
I read a bit lmql, seems to be doing pretty much what I would like to avoid, knowing the model internals (here the tokenizer) to control the output via input_id biases.
Even worse, you usually don't want the same biases depending on where you are in the generation.

This seems a bit higher level and closer to something which would make sense IMO: https://github.com/guidance-ai/guidance
I don't think we want to expose anything with regard to model internals. It makes things quite hard to use.

Still a lot of work when we're not working with a single model, and a bunch, which use different tokenizers with different properties. But the higher level the control, the easier it is for us to maintain ideally.

**Comment by MarcusDunn** [comment]
Just to clarify - the PR as it currently is takes in strings to bias towards, not tokens. The caller does not need to know token_ids and can pass in strings larger than one token.

**Comment by MarcusDunn** [comment]
Just to clarify - the PR as it currently is takes in strings to bias towards, not tokens. The caller does not need to know token_ids and can pass in strings larger than one token.

Is something like guidance within the scope for tgi?

Where could I find the roadmap you mentioned?

**Comment by charles-dyfis-net** [comment]
> I read a bit lmql, seems to be doing pretty much what I would like to avoid, knowing the model internals (here the tokenizer) to control the output via input_id biases.

That's not a burden LMQL puts on its users, but one it takes on itself -- much as Guidance does.

> Even worse, you usually don't want the same biases depending on where you are in the generation.

Correct, and that's why LMQL breaks down an operation into smaller requests to the model, again, as Guidance does.

**Comment by Narsil** [comment]
@charles-dyfis-net . But that means stopping and restarting queries all the time. Meaning forgetting KV all the time pretty much.

This seems like a very high cost latency wise.
Keeping KV around hoping for client to reconnect is also not great. (How long do you keep it around, how do you cache hit, etc..).

**Comment by charles-dyfis-net** [comment]
> @charles-dyfis-net . But that means stopping and restarting queries all the time. Meaning forgetting KV all the time pretty much.
> 
> This seems like a very high cost latency wise. Keeping KV around hoping for client to reconnect is also not great. (How long do you keep it around, how do you cache hit, etc..).

So when you describe guidance as something that you want to support, are you saying you'd want to run it server-side, built into TGI? If you run Guidance server-side... well, you could run LMQL queries server-side as well, though that's going to leave a lot of users (like ones who want to implement tool-calling workflows) dissatisfied. And unless y'all have a route to Microsoft's team that isn't open to the public, I'd worry about maintainer accessibility; one of the significant factors behind my decision to use LMQL over Guidance is the availability of a highly responsive and accessible upstream.

One way to split the difference is to divide into multiple operations whenever tool use or something else that needs to happen on the remote end takes place; LMQL upstream has been toying with the idea of supporting that explicitly.

With respect to the KV cache -- personally, I'd bind it to a specific [LMTP](https://github.com/eth-sri/lmql/blob/main/src/lmql/models/lmtp/README.md)-over-websocket connection, since a given LMQL client will be reusing a single transport over time in nontrivial cases. (The interactive debugger/playground creates a new transport instance per call, but it's not typical of more serious applications in that respect).

**Comment by MarcusDunn** [comment]
Looks like we'll be getting one better than this PR.
 https://github.com/huggingface/text-generation-inference/issues/505#issuecomment-1708367609

I'll close for now.
